lib/spack/spack/modules.py

```diff
@@ -159,43 +243,61 @@ def write(self):
         # installation prefix
         env = inspect_path(self.spec.prefix)
 
-        # Let the extendee modify their extensions before asking for
+        # Let the extendee/dependency modify their extensions/dependencies before asking for
         # package-specific modifications
         spack_env = EnvironmentModifications()
-        for item in self.pkg.extendees:
-            try:
-                package = self.spec[item].package
-                package.setup_dependent_package(self.pkg.module, self.spec)
-                package.setup_dependent_environment(spack_env, env, self.spec)
-            except:
-                # The extends was conditional, so it doesn't count here
-                # eg: extends('python', when='+python')
-                pass
+        # TODO : the code down below is quite similar to build_environment.setup_package and needs to be
+        # TODO : factored out to a single place
+        for item in dependencies(self.spec, 'All'):
```

Comment:
@tgamblin Adding these three lines:
```
for item in self.spec.traverse(order='post', depth=True, cover='nodes', root=False):
    tty.msg(item)
tty.msg('')
```
immediately above here gives to me the following output for `gcc@5.3.0` during `spack module refresh`:
```
==> (1, binutils@2.26%gcc@4.8+gold~krellpatch~libiberty=production)
==> (1, gmp@6.1.0%gcc@4.8=production)
==> (2, gmp@6.1.0%gcc@4.8=production)
==> (1, isl@0.14%gcc@4.8=production^gmp@6.1.0%gcc@4.8=production)
==> (2, gmp@6.1.0%gcc@4.8=production)
==> (3, gmp@6.1.0%gcc@4.8=production)
==> (2, mpfr@3.1.4%gcc@4.8=production^gmp@6.1.0%gcc@4.8=production)
==> (1, mpc@1.0.3%gcc@4.8=production^gmp@6.1.0%gcc@4.8=production^mpfr@3.1.4%gcc@4.8=production)
==> (2, gmp@6.1.0%gcc@4.8=production)
==> (1, mpfr@3.1.4%gcc@4.8=production^gmp@6.1.0%gcc@4.8=production)
```
The point that I want you to note is that we are visiting `gmp@6.1.0%gcc@4.8=production` multiple times. The behavior won't change if you remove `depth=True` and is not consistent with what happens at build time (where the nodes are visited only once). 

I really don't get if we miss a call to some method before entering here (to collapse nodes that refers to the same spec) or if it is the way this is supposed to work.